connectors/mongo/data_op_processor.go

```diff
@@ -39,7 +40,111 @@ func (mc *BaseMongoConnector) HandleBarrierMessage(barrierMsg iface.DataMessage)
 	return nil
 }
 
-func (mc *BaseMongoConnector) ProcessDataMessage(dataMsg iface.DataMessage) error {
+type dataMsgIdIndex struct {
+	dataMsgId []byte
+	index     int
+}
+
+// returns the new item or existing item, and whether or not a new item was added
+func addToIdIndexMap(m map[int][]*dataMsgIdIndex, dataMsg iface.DataMessage) (*dataMsgIdIndex, bool) {
+	h := hashDataMsgId(dataMsg)
+	items, found := m[h]
+	if found {
+		for _, item := range items {
+			if slices.Equal(item.dataMsgId, *dataMsg.Id) {
+				return item, false
+			}
+		}
+	}
+	item := &dataMsgIdIndex{*dataMsg.Id, -1}
+	m[h] = append(items, item)
+	return item, true
+}
+
+// ProcesDataMessages assumes all have the same Database and Collection
+func (mc *BaseMongoConnector) ProcessDataMessages(dataMsgs []iface.DataMessage) error {
+	if len(dataMsgs) == 0 {
+		return nil
+	}
+	dataMsg := dataMsgs[0]
+	dbName := dataMsg.Loc.Database
+	colName := dataMsg.Loc.Collection
+	maxSeqNum := max(dataMsg.SeqNum, mc.Status.WriteLSN)
+
+	collection := mc.Client.Database(dbName).Collection(colName)
+
+	var models []mongo.WriteModel
+	// keeps track of the index in models for a particular document because last write wins
+	hashToDataMsgIdIndex := map[int][]*dataMsgIdIndex{}
+
+	for _, dataMsg := range dataMsgs {
+		idType := bsontype.Type(dataMsg.IdType)
+		maxSeqNum = max(maxSeqNum, dataMsg.SeqNum)
+
+		switch dataMsg.MutationType {
+		case iface.MutationType_Insert, iface.MutationType_Update:
+			dmii, isNew := addToIdIndexMap(hashToDataMsgIdIndex, dataMsg)
```

Comment:
Maybe something more descriptive instead of "dmii" ?

Reply:
The usage is pretty localized, dmii is just short for dataMessageIdIndex, so I think this should be fine to keep.